### PR TITLE
Allow auto creating new terms or selecting existing ones

### PR DESCRIPTION
**Pull request by amitaibu**
- Fix taxonomy term integration
- Integrate with select2 on the Angular, so you can select existing terms and create new ones as-well
- Node and terms are created in a _single_ request!

Needs some more cleanup, but already working!
angularjs_form___site-install

### CODE REVIEW

**Comment by amitaibu** [comment]
Ready for review.

File: modules/restful_angular_example/plugins/restful/taxonomy_term/tags/1.0/RestfulEntityTaxonomyTermTags.class.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  /**
+   * Overrides \RestfulEntityBaseTaxonomyTerm::createEntity().
+   *
+   * Allow access to create "Tags" resource for privileged users, as
+   * we can't use entity_access() since entity_metadata_taxonomy_access()
+   * denies it for a non-admin user.
+   */
+  public function createEntity() {
+    $account = $this->getAccount();
+    if (!user_access('create article content', $account)) {
```
Comment (e0ipso) [inline modules/restful_angular_example/plugins/restful/taxonomy_term/tags/1.0/RestfulEntityTaxonomyTermTags.class.php]: What about overriding `\RestfulEntityBaseTaxonomyTerm::access()` for this?

File: modules/restful_angular_example/plugins/restful/taxonomy_term/tags/1.0/RestfulEntityTaxonomyTermTags.class.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    $entity = entity_create($this->entityType, $values);
+
+    $wrapper = entity_metadata_wrapper($this->entityType, $entity);
+
+    $this->setPropertyValues($wrapper);
+    return $this->viewEntity($wrapper->getIdentifier());
+  }
+
+  /**
+   * Overrides \RestfulEntityBaseTaxonomyTerm::createEntity().
```
Comment (e0ipso) [inline modules/restful_angular_example/plugins/restful/taxonomy_term/tags/1.0/RestfulEntityTaxonomyTermTags.class.php]: c&p error

``` php
/**
 * Overrides \RestfulEntity…::checkPropertyAccess().
 */
```

File: modules/restful_angular_example/plugins/restful/taxonomy_term/tags/1.0/RestfulEntityTaxonomyTermTags.class.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+   *
+   * Allow user to create a label for the term.
```
Comment (e0ipso) [inline modules/restful_angular_example/plugins/restful/taxonomy_term/tags/1.0/RestfulEntityTaxonomyTermTags.class.php]: Maybe:
`Allow user to create a label for the tag.`?

File: modules/restful_angular_example/restful_angular_example.module [hunk 4]
```diff
@@ -73,4 +86,4 @@ function restful_angular_example_form_page() {
   // view.
   $url = url($app_path . '/views/main.html', array('absolute' => TRUE));
   return theme('restful_angular_example_angular_form', array('url' => $url));
-}
+}
```
Comment (e0ipso) [inline modules/restful_angular_example/restful_angular_example.module]: Missing line break.

File: plugins/restful/RestfulEntityBase.php [hunk 5]
```diff
@@ -614,6 +618,37 @@ public function getQueryForList() {
...
+   * @return array
+   *   Array with the found terms keys by the entity ID.
+   *   ID. Otherwise, if the field allows auto-creating tags, the ID will be the
+   *   term name, to indicate for client it is an unsaved term.
+   *
+   * @see taxonomy_autocomplete()
+   */
+  protected function getListByAutocomplete() {
+    $request = $this->getRequest();
+    if (empty($request['string'])) {
```
Comment (e0ipso) [inline plugins/restful/RestfulEntityBase.php]: `$request['string']` seems pretty generic, what about `$request['autocomplete']['string']`. This way we can also pass an operator for the SQL query being `contains|starts_with|ends_with` (defaulting it to `contains`).

File: plugins/restful/RestfulEntityBase.php [hunk 6]
```diff
@@ -614,6 +618,37 @@ public function getQueryForList() {
...
+   * Return the values of the types tags, with the ID.
+   *
+   * @return array
+   *   Array with the found terms keys by the entity ID.
+   *   ID. Otherwise, if the field allows auto-creating tags, the ID will be the
+   *   term name, to indicate for client it is an unsaved term.
+   *
+   * @see taxonomy_autocomplete()
+   */
+  protected function getListByAutocomplete() {
```
Comment (e0ipso) [inline plugins/restful/RestfulEntityBase.php]: `getListForAutocomplete`?

File: plugins/restful/RestfulEntityBaseTaxonomyTerm.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    $query = db_select('taxonomy_term_data', 't');
+    $query->addTag('translatable');
+    $query->addTag('term_access');
+
+    // Select rows that match by term name.
+    return $query
+      ->fields('t', array('tid', 'name'))
+      ->condition('t.vid', $bundles, 'IN')
+      ->condition('t.name', '%' . db_like($string) . '%', 'LIKE')
```
Comment (e0ipso) [inline plugins/restful/RestfulEntityBaseTaxonomyTerm.php]: Here we would use the aforementioned autocomplete operator.

**Comment by e0ipso** [comment]
Should we have tests for taxonomy terms? Follow up?

**Comment by amitaibu** [comment]
I've generalized the auto complete to work on all entities, and added tests for taxonomy integration and autocomplete.

We can tackle any remaining issues in a follow up. Merged.

**Pull request closed without merging**